### DISCUSSION TITLE
coredns

### DISCUSSION THREAD

**Discussion by bpopovich44** [rancher/rke2#3493] [General]
I am using rke2 1.24.7 and the dns works fine for a while, but then seems to bottleneck with coredns giving the following errors.  It still works, but is intermittent.  Ideas?

[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.899644677s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.899697325s
[WARNING] plugin/health: Local health request to "http://:8080/health" failed: Get "http://:8080/health": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 3.200311788s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.000010172s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.374516495s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.300175671s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.100442093s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.799865368s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.500314764s
[WARNING] plugin/health: Local health request to "http://:8080/health" failed: Get "http://:8080/health": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.600244392s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.671504106s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.585956884s
[WARNING] plugin/health: Local health request to "http://:8080/health" failed: Get "http://:8080/health": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.003437761s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.999504985s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.100054141s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.100353923s
[WARNING] plugin/health: Local health request to "http://:8080/health" failed: Get "http://:8080/health": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.300393019s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.499457076s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 2.400130227s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.500350949s
[WARNING] plugin/health: Local health request to "http://:8080/health" failed: Get "http://:8080/health": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.59925706s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.399565328s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.10005273s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.300457729s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.504547838s
[WARNING] plugin/health: Local health request to "http://:8080/health" took more than 1s: 1.701145327s

**Comment by brandond** [comment]
How much CPU time is the coredns pod using? What does `kubectl top pod -n kube-system` show? If you describe the coredns pod, is it being restarted for any reason?

**Reply by bpopovich44** [reply]
Thank you...

```
kubectl top pod -n kube-system
NAME CPU(cores) MEMORY(bytes)
cloud-controller-manager-rke2a-ha 3m 25Mi
cloud-controller-manager-rke2b-ha 2m 23Mi
cloud-controller-manager-rke2c-ha 1m 20Mi
etcd-rke2a-ha 59m 275Mi
etcd-rke2b-ha 37m 271Mi
etcd-rke2c-ha 27m 270Mi
kube-apiserver-rke2a-ha 108m 1488Mi
kube-apiserver-rke2b-ha 83m 1515Mi
kube-apiserver-rke2c-ha 66m 1231Mi
kube-controller-manager-rke2a-ha 14m 165Mi
kube-controller-manager-rke2b-ha 2m 37Mi
kube-controller-manager-rke2c-ha 2m 33Mi
kube-proxy-rke00 1m 38Mi
kube-proxy-rke02 18m 51Mi
kube-proxy-rke2a-ha 1m 26Mi
kube-proxy-rke2b-ha 1m 29Mi
kube-proxy-rke2c-ha 1m 24Mi
kube-scheduler-rke2a-ha 3m 49Mi
kube-scheduler-rke2b-ha 2m 53Mi
kube-scheduler-rke2c-ha 2m 45Mi
kube-vip-ds-2bgrz 1m 21Mi
kube-vip-ds-phdfn 1m 20Mi
kube-vip-ds-sfttf 4m 22Mi
rke2-canal-5bf8v 25m 298Mi
rke2-canal-k2cfp 20m 168Mi
rke2-canal-rcvp5 17m 159Mi
rke2-canal-rx8mz 24m 430Mi
rke2-canal-z22bt 19m 160Mi
rke2-coredns-rke2-coredns-58fd75f64b-4592b 81m 83Mi
rke2-coredns-rke2-coredns-58fd75f64b-shrld 57m 31Mi
rke2-coredns-rke2-coredns-58fd75f64b-txgws 58m 30Mi
rke2-coredns-rke2-coredns-autoscaler-768bfc5985-cr26k 1m 12Mi
rke2-ingress-nginx-controller-5wqcz 2m 125Mi
rke2-ingress-nginx-controller-8bhzm 21m 3482Mi
rke2-ingress-nginx-controller-g5jmp 16m 2989Mi
rke2-ingress-nginx-controller-r2g4b 2m 132Mi
rke2-ingress-nginx-controller-tpph6 3m 129Mi
rke2-metrics-server-67697454f8-9whhh 3m 35Mi
```

There are no restarts except I did one this morning:
```
IP: 10.42.3.9
Controlled By: ReplicaSet/rke2-coredns-rke2-coredns-58fd75f64b
Containers:
coredns:
Container ID: containerd://1b2238f5330490e8d1654fe08a61dc84d678dd117a2bb84349e9615da81254fd
Image: rancher/hardened-coredns:v1.9.3-build20221011
Image ID: docker.io/rancher/hardened-coredns@sha256:3020505d755499eb8d22f631aa88fc597b4849eae18404401087669a6bc5ca06
Ports: 53/UDP, 53/TCP
Host Ports: 0/UDP, 0/TCP
Args:
-conf
/etc/coredns/Corefile
State: Running
Started: Tue, 25 Oct 2022 04:31:37 -0600
Last State: Terminated
Reason: Unknown
Exit Code: 255
Started: Tue, 25 Oct 2022 03:22:33 -0600
Finished: Tue, 25 Oct 2022 04:31:25 -0600
Ready: True
Restart Count: 1
Limits:
cpu: 100m
memory: 128Mi
Requests:
cpu: 100m
memory: 128Mi
Liveness: http-get http://:8080/health delay=60s timeout=5s period=10s #success=1 #failure=5
Readiness: http-get http://:8181/ready delay=30s timeout=5s period=10s #success=1 #failure=5
Environment:
```

**Reply by bpopovich44** [reply]
Am I reading it right from the describe command it maxes at 100m? Maybe it is maxing out cpu usage for the pod?

Now it is at 100m. I'm not sure if it caps and stops/fails or holds there, or if this isnt related at all
rke2-coredns-rke2-coredns-58fd75f64b-4592b 100m 109Mi
rke2-coredns-rke2-coredns-58fd75f64b-shrld 37m 30Mi
rke2-coredns-rke2-coredns-58fd75f64b-txgws 39m 28Mi

**Reply by brandond** [reply]
yes, it is by default limited to 1/10th of a core. This is usually more than sufficient. If you have a workload that is for some reason particularly hard on coredns, you may need to customize the rke2-coredns chart configuration to raise the requests/limits.

https://github.com/rancher/rke2-charts/blob/main/charts/rke2-coredns/rke2-coredns/1.19.401/values.yaml#L18-L21

**Reply by brandond** [reply]
I would probably also investigate what you have that's hammering so hard on coredns though. It shouldn't use that much, unless something is making a TON of dns requests.

**Reply by bpopovich44** [reply]
Wow.  Thank you, I really appreciate your help.  Two questions:
1) Should I replace the resources and requests or just resources?
resources:
  limits:
    cpu: 100m
    memory: 128Mi
  requests:
    cpu: 100m
    memory: 128Mi

2) I cannot find the location of this yaml, even with the find command.  It is not in the /var/lib/rancher/server/manifest file and if I search for coredns.yaml it is not present either

**Reply by bpopovich44** [reply]
Yea, the requests are coming from gitlab runners and I can see them making so many requests

**Reply by brandond** [reply]
You would want to customize the rke2-coredns chart values via a HelmChartConfig, as covered in the documentation at https://docs.rke2.io/helm/#customizing-packaged-components-with-helmchartconfig

**Reply by bpopovich44** [reply]
Brandon, thank you for your help...
Will I apply this with kubectl apply -f rke2-coredns-config.yaml or do I need to run a helm install?

**Reply by brandond** [reply]
You would apply it to the cluster. A helm upgrade operation will be triggered by the configuration change.

**Reply by bpopovich44** [reply]
And if I apply this yaml, will it overwrite current settings for coredns?

**Reply by brandond** [reply]
Did you make manual changes to the coredns configuration? If so that's a bad idea to begin with, as the changes would be lost when we update coredns in future releases.

**Reply by bpopovich44** [reply]
I'm a bit confused.  In the second link it says I need to make sure the name and namespace match, but they aren't in the config from the first link.  Do I need to add those parameters?  Or is the yaml going to know to just make the updates in the correct namespace?  The yaml also asks for service account.  Thank you for your help, I want to make sure I get it correct the first time.

**Reply by bpopovich44** [reply]
Do I need to un-comment the auto scaling too for rke2-coredns-rke2-coredns-autoscaler

**Reply by brandond** [reply]
I'm not sure exactly what you're doing. Where are you missing a namespace? What is asking for a service account? All you need to customize is the resources in the chart values. Instead of setting the image and tag as in the example, set the resources.

Find the bits of the values.yaml you want to change, and put that in the HelmChartConfig valuesContent section. Make sure you get the indentation correct.

```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-coredns
  namespace: kube-system
spec:
  valuesContent: |-
    resources:
      limits:
        cpu: 200m
        memory: 256Mi
      requests:
        cpu: 100m
        memory: 128Mi
```

**Reply by bpopovich44** [reply]
Hi Brandon,
Thanks for you help.  I made the changes you suggested and it worked great.  I had to make a few changes as we went up to almost 400m cpu.  Should I install NodeLocal DNSCache to help or do you have any ideas?  The gitlab pods are making alot of calls and I'm sure they are repetitive.  I read that default cache it about a minute and some of the runners run for more than 30 min.

apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-coredns
  namespace: kube-system
spec:
  valuesContent: |-
    nodelocal:
      enabled: true

**Reply by brandond** [reply]
Yeah, that would probably be helpful.

**Reply by bpopovich44** [reply]
Hi Brandon,
We still get several DNS issues.  The coredns logs dont show any errors, but when gobs run we get fluctuations when they work, but then a bunch wont work.  Also getting port 443 errors

**Reply by bpopovich44** [reply]
Here is an example:
Created fresh repository.
fatal: unable to access 'https://gitlab.seakr.com/
build.git/': Could not resolve host: gitlab.seakr.com

**Reply by bpopovich44** [reply]
ssh: Could not resolve hostname gitlab.seakr.com: Temporary failure in name resolution

**Reply by bpopovich44** [reply]
[WARNING] plugin/kubernetes: Kubernetes API connection failure: Get "https://10.43.0.1:443/version": dial tcp 10.43.0.1:443: i/o timeout

**Reply by bpopovich44** [reply]
Hi Brandon, 
Not sure if this is still open but still getting errors with DNS, but the logs are empty.  I put in another request for help because not sure if this is still open.  
Thank you

**Reply by brandond** [reply]
This is a discussion thread, it doesn't ever really close, although you can mark a comment as the answer. If you have a Suse/Rancher support contract, you can open a ticket - but this is just best-effort community assistance, and we don't always have time to respond.

I don't have any immediate suggestions for you other than checking for errors in other logs and ensuring that the node-local dns pods aren't also being throttled.

**Reply by bpopovich44** [reply]
No worries on the response, I know you have other stuff going on, but I appreciate your help.
I ended up deleting the NodeLocal DNS because it seemed to make things worse.  I also noticed it didn't launch a pod which I believe it is supposed to.

**Reply by brandond** [reply]
It should result in creation of a daemonset that runs a coredns pod on each of the nodes. I do suspect that it might help with your load issue.

**Reply by bpopovich44** [reply]
Hey!  So it turns out there was a bad NIC in one of the servers causing all the issues.  I drained the node and pushed coredns to another worker node and it fixed the issues.  Although the CPUs for coredns are still hitting about 250m, you definitely helped fix that issue so thank you.  The coredns logs are empty.  I really appreciate your help.
